### PR TITLE
fix: memberlist delete operation

### PR DESCRIPTION
**Pull request by NickAnge**
**What this PR does**:

After the rollout of delete operation in memberlist we observed 2 things:
- Error `failed to decode value: snappy: corrupt input`. ([logs](https://ops.grafana-ops.net/explore?schemaVersion=1&panes=%7B%22ygx%22%3A%7B%22datasource%22%3A%22--+Mixed+--%22%2C%22queries%22%3A%5B%7B%22refId%22%3A%22A%22%2C%22expr%22%3A%22sum+by%28namespace%29%28count_over_time%28%7Bcluster%3D%5C%22dev-us-central-0%5C%22%2Cnamespace%3D%5C%22mimir-dev-11%5C%22%2Ccontainer%3D%5C%22distributor%5C%22%7D++%7C%3D+%60deleted+old+replica%60++%5B%24__auto%5D%29%29%22%2C%22range%22%3Atrue%2C%22datasource%22%3A%7B%22type%22%3A%22loki%22%2C%22uid%22%3A%22OP27Xzxnk%22%7D%2C%22editorMode%22%3A%22code%22%2C%22queryType%22%3A%22range%22%2C%22direction%22%3A%22backward%22%2C%22hide%22%3Atrue%7D%2C%7B%22refId%22%3A%22B%22%2C%22expr%22%3A%22%7Bcluster%3D%5C%22dev-us-central-0%5C%22%2C+namespace%3D%5C%22mimir-dev-11%5C%22%2C+container%3D%5C%22distributor%5C%22%7D+%7C+logfmt+%7C+detected_level%3D%5C%22error%5C%22+%7C%3D+%5C%22failed+to+store+received+value%5C%22%22%2C%22queryType%22%3A%22range%22%2C%22datasource%22%3A%7B%22type%22%3A%22loki%22%2C%22uid%22%3A%22OP27Xzxnk%22%7D%2C%22editorMode%22%3A%22code%22%2C%22direction%22%3A%22backward%22%2C%22hide%22%3Afalse%7D%5D%2C%22range%22%3A%7B%22from%22%3A%221738594922337%22%2C%22to%22%3A%221738601933484%22%7D%2C%22panelsState%22%3A%7B%22logs%22%3A%7B%22visualisationType%22%3A%22logs%22%7D%7D%7D%7D&orgId=1))
- Delete was happening but the actual entry was not deleted completed. So the entry is deleted it and then comes back. 

In this PR we tackle these two issues by:
- returning an actual `change` when the only thing that gets updated is the Delete status, so we do not have to `handlePossibleNilEncode` which was returning an []byte{} that was causing the Decode issue.
- returning no change when the incoming entry has Delete true. Thist was an edge case in which the current HA tracker had deleted the entry and another one was coming with delete True. This can happen if we have regular push/pull interval.

Verified it locally. 

- [ ] Need to test it in a dev-cell. mimir-dev-11, would be the cell. https://github.com/grafana/deployment_tools/pull/220742/files

### Tests
- Introduces `checkMemberlistEntry`, that will constantly checks the memberlist entries up to specific amount of time before considers the test failure. It utilizes test.poll
- extend `TestDelete` to check if there are any encoding errors during the Delete flow. Covers `failed to decode value: snappy: corrupt input`

**Which issue(s) this PR fixes**:

Fixes #<issue number>

**Checklist**
- [x] Tests updated
- [ ] `CHANGELOG.md` updated - the order of entries should be `[CHANGE]`, `[FEATURE]`, `[ENHANCEMENT]`, `[BUGFIX]`

### CODE REVIEW

**Review by dimitarvdimitrov — COMMENTED** [review]
the change makes sense, but it definitely needs tests. If you can even cover cases which haven't been covered so far, that'd be great

File: kv/memberlist/memberlist_client.go [hunk 1]
```diff
@@ -1695,19 +1702,3 @@ func updateTimeMillis(ts time.Time) int64 {
 	}
 	return ts.UnixMilli()
 }
-
-func handlePossibleNilEncode(codec codec.Codec, change Mergeable) ([]byte, error) {
```
Comment (NickAnge) [inline kv/memberlist/memberlist_client.go:1699]: I decided that the `handle*` funcions are not needed since we handle the change=nil at the ` mergeValueForKey` level. That was the main reason these were needed at the first place. I validated that the `[]byte{}`was the one causing the `Error failed to decode value: snappy: corrupt input` at the first place. Please let me know what you think

**Comment by NickAnge** [comment]
Update after the test in mimir-dev-11
- The entry is deleted as expected ✅ 

Bad news
- We still  `see failed to decode value: snappy: corrupt input` ([logs](https://ops.grafana-ops.net/explore?schemaVersion=1&panes=%7B%22ygx%22:%7B%22datasource%22:%22--%20Mixed%20--%22,%22queries%22:%5B%7B%22refId%22:%22A%22,%22expr%22:%22sum%20by%28namespace%29%28count_over_time%28%7Bcluster%3D%5C%22dev-us-central-0%5C%22,namespace%3D%5C%22mimir-dev-11%5C%22,container%3D%5C%22distributor%5C%22%7D%20%20%7C%3D%20%60deleted%20old%20replica%60%20%20%5B$__auto%5D%29%29%22,%22range%22:true,%22datasource%22:%7B%22type%22:%22loki%22,%22uid%22:%22OP27Xzxnk%22%7D,%22editorMode%22:%22code%22,%22queryType%22:%22range%22,%22direction%22:%22backward%22,%22hide%22:true%7D,%7B%22refId%22:%22B%22,%22expr%22:%22%7Bcluster%3D%5C%22dev-us-central-0%5C%22,%20namespace%3D%5C%22mimir-dev-11%5C%22,%20container%3D%5C%22distributor%5C%22%7D%20%7C%20logfmt%20%7C%20detected_level%3D%5C%22error%5C%22%20%7C%3D%20%5C%22failed%20to%20store%20received%20value%5C%22%22,%22queryType%22:%22range%22,%22datasource%22:%7B%22type%22:%22loki%22,%22uid%22:%22OP27Xzxnk%22%7D,%22editorMode%22:%22code%22,%22direction%22:%22backward%22,%22hide%22:false%7D%5D,%22range%22:%7B%22from%22:%221739267302260%22,%22to%22:%221739269572211%22%7D,%22panelsState%22:%7B%22logs%22:%7B%22visualisationType%22:%22logs%22%7D%7D%7D%7D&orgId=1)) after the delete operation , which I am planning to investigate

**Comment by NickAnge** [comment]
Closing this PR, we merged the changes with https://github.com/grafana/dskit/pull/645/files. Which solves also the `see failed to decode value: snappy: corrupt input` Tested in mimir-dev-14
